pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go

```diff
@@ -0,0 +1,159 @@
+		queue:             queue.NewQueue(time.Second),
+		clusterLocalHosts: clusterLocalHosts,
+	}
+
+	serviceExportController.serviceInformer = kubeClient.KubeInformer().Core().V1().Services().Informer()
+	serviceExportController.serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			serviceExportController.queue.Push(func() error {
+				serviceObj, err := convertToService(obj)
+				if err != nil {
+					return err
+				}
+				return serviceExportController.HandleNewService(serviceObj)
+			})
+		},
+		DeleteFunc: func(obj interface{}) {
+			serviceExportController.queue.Push(func() error {
+				serviceObj, err := convertToService(obj)
+				if err != nil {
+					return err
+				}
+				return serviceExportController.HandleDeletedService(serviceObj)
+			})
+		},
+	})
+
+	return serviceExportController, nil
+}
+
+func (sc *ServiceExportController) Run(stopCh <-chan struct{}) {
+	log.Infof("Syncing existing services and serviceexports...")
+	sc.doInitialFullSync()
+	log.Infof("serviceexport sync complete")
+	cache.WaitForCacheSync(stopCh, sc.serviceInformer.HasSynced)
+	log.Infof("ServiceExport controller started")
+	go sc.queue.Run(stopCh)
+}
+
+func (sc *ServiceExportController) HandleNewService(obj *v1.Service) error {
+	if sc.isServiceClusterLocal(obj) {
+		return nil //don't don anything for marked clusterlocal services
+	}
+	return sc.createServiceExportIfNotPresent(obj)
+}
+
+func (sc *ServiceExportController) HandleDeletedService(obj *v1.Service) error {
+	return sc.deleteServiceExportIfPresent(obj)
+}
+
+func convertToService(obj interface{}) (*v1.Service, error) {
+	cm, ok := obj.(*v1.Service)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			return nil, fmt.Errorf("couldn't get object from tombstone %#v", obj)
+		}
+		cm, ok = tombstone.Obj.(*v1.Service)
+		if !ok {
+			return nil, fmt.Errorf("tombstone contained object that is not a Service %#v", obj)
+		}
+	}
+	return cm, nil
+}
+
+func (sc *ServiceExportController) isServiceClusterLocal(service *v1.Service) bool {
+	for _, host := range sc.clusterLocalHosts {
+		hostComponents := strings.Split(host, ".")
+		//name match
+		if (hostComponents[0] == "*" || hostComponents[0] == service.Name) && (hostComponents[1] == "*" || hostComponents[1] == service.Namespace) {
+			return true
+		}
+	}
+	return false
+}
+
+func (sc *ServiceExportController) createServiceExportIfNotPresent(service *v1.Service) error {
+	serviceExport := v1alpha1.ServiceExport{}
+	serviceExport.Namespace = service.Namespace
+	serviceExport.Name = service.Name
+
+	_, err := sc.client.MulticlusterV1alpha1().ServiceExports(service.Namespace).Create(context.TODO(), &serviceExport, metav1.CreateOptions{})
+
+	if err != nil && strings.Contains(err.Error(), "already exists") {
+		err = nil //This is the error thrown by the client if there is already an object with the name in the namespace. If that's true, we do nothing
```

Comment:
what if the contents are wrong?

Reply:
For this implementation, we are making the intentional choice that we don't care, and don't support cases in which some other entity (either human or 3rd party MCS controller) modifies a ServiceExport. So if it exists, we stop processing (to minimize cases where the other entity overwrite may be desired behavior).